### DISCUSSION TITLE
FileUpload file not saved in the specified final directory  ->directory()

### DISCUSSION THREAD

**Discussion by klytron** [filamentphp/filament#9465] [General]
Uploaded file is deleted in the folder "storage/app/livewire-tmp" but does not exist in the destination "storage/app/uploaded-csv" the destination folder is created if it does not exist, but the uploaded file is not saved.

I created a non-resource page to just upload files into a folder.

php artisan make:filament-page ImportGalleries --type=custom

````

namespace App\Filament\Pages;

use Filament\Actions\Action;
use Filament\Pages\Page;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Forms\Contracts\HasForms;
use Filament\Forms\Form;
use Illuminate\Contracts\View\View;
use Filament\Forms\Components\FileUpload;
use Filament\Notifications\Notification;

class ImportGalleries extends Page implements HasForms
{
    use InteractsWithForms;

    protected static ?string $navigationIcon = 'heroicon-o-document-text';

    protected static string $view = 'filament.pages.import-galleries';

    public ?array $data = [];

    public function mount(): void
    {
        $this->form->fill();
    }

    public function form(Form $form): Form
    {
        return $form
            ->schema([
                FileUpload::make('csv_file')
                    ->acceptedFileTypes([
                            'text/csv'
                        ])
                    ->preserveFilenames()
                    ->multiple()
                    ->hint('Max file size 200MB')
                    ->label('Gallery CSV File(s)')
                    ->maxSize(212288)
                    ->disk('local')  // storage/app
                    ->directory('uploaded-csv')  // storage/app/uploaded-csv
                    ->visibility('private')
                    ->required()
                    ->moveFiles(), //move from temp folder, instead of copying. helps to save disk space.
            ])
            ->statePath('data');
    }

    public function submitForm(): void
    {
        $attrs = $this->form->getState();

        Notification::make()
            ->success()
            ->title('file uploaded')
            ->send();
    }

    public function getFormActions(): array
    {
        return [
            Action::make('save')
                ->icon('heroicon-m-clipboard')
                ->submit('save'),
        ];
    }

}
`````
The view file:

`````
<x-filament-panels::page>

    <p> 
        After uploading allow up to 20 minutes for the background processor to import your galleries
        for display.
    </p>

    <x-filament-panels::form wire:submit="submitForm">

        {{ $this->form }}
        
        <x-filament-panels::form.actions 
            :actions="$this->getFormActions()"
        />
    
        <x-filament-actions::modals />

    </x-filament-panels::form>


</x-filament-panels::page>

```````

Uploaded file should be moved from "storage/app/livewire-tmp"  and saved at the destination "storage/app/uploaded-csv" 

filament/forms Version
v3.0.90

Laravel Version
v10.30.1

Livewire Version
v3.1.0

PHP Version
PHP 8.2.0

**Comment by sacalito** [comment]
same problem here

**Comment by paulruderman** [comment]
Same problem here, only in production using storage buckets.  File appears correctly in the temporary directory but does not get moved to the permanent location.  The permanent directory is created if it's missing, but the file is not moved into place.

**Comment by basherdesigns** [comment]
I'm having very similar issue as well. 

Laravel v11.9.2 / FilamentPHP v.v3.2.85 / Spatie Media Library

All functionality works correctly on dev environment using CloudFlare R2 as storage. Moved to Production and when I add an image, it uploads to R2 bucket temp directory /livewire-tmp/ymWaqsRePMGAOC44Qik8veBKZJvozU-metaaW1hZ2UtYXNzZXQuanBn-.jpg (for example) Then hit Save on form and get 500 Error.

Log shows:

```
production.ERROR: Illuminate\Translation\PotentiallyTranslatedString::__toString(): Return value must be of type string, null returned {"userId":2,"exception":"[object] (TypeError(code: 0): Illuminate\Translation\PotentiallyTranslatedString::__toString(): Return value must be of type string, null returned at /home/forge/domain.com/vendor/laravel/framework/src/Illuminate/Translation/PotentiallyTranslatedString.php:89)
[stacktrace]
```
Image is not uploaded from temp directory to r2 public bucket and record is not saved in the media table for the image on the model.

Here is my form element:

```
SpatieMediaLibraryFileUpload::make('location-images')
                ->imageEditor()
                ->disk('r2images')
                ->label('Location Photos')
                ->columnSpanFull()
                ->collection('location-images')
                ->multiple()
                ->image()
                ->reorderable(),
```

If I remove the image() function, the model saves with Save notification but image is not saved to correct bucket and not added to the media table.

**Reply by basherdesigns** [reply]
I should also add, when hitting Save on another Model form after the image is uploaded to livewire-tmp, the record is not saved and validation error shows 

"The <fieldname> field must be a file of type: image/*"

I've tried .jpg and .png with no success. Works fine in dev using the same storage buckets.

**Comment by heruprambadi** [comment]
in my case, just use "php artisan storage:link" and it solved.

**Comment by jubox974** [comment]
I'm facing the same issue here FileUpload move to the temporary folder in bucket but not in the final directory after save. 

Any solutions ?

**Comment by basherdesigns** [comment]
Anyone find a solution for this issue?  My app is hosted on a Laravel Forge / Digital Ocean server which has 2 apps hosted. I've yet to get a resolution for this!  What are we missing...........

**Comment by TimeZHero** [comment]
I had the same issue using a remote filesystem (S3) and I found out what was happening.

## Why no errors?
First of all, there is a _silent_ fail due to the filesystem having `throw => false`. 
This allows the code to handle a unsuccessful upload, but in this case Livewire / Filament are not handling it, making debugging really difficult.

Setting the `throw => true` will reveal what is happening, with a `400 Bad Request` towards the bucket:
The error further explains that the issue is related to the bucket **NOT** having `ACL` enabled. 

## The actual issue
This happens because the default `->visibility()` in the `FileUpload` field is `public`, and Filament uses this param to decide between the methods `storePubliclyAs` and `storeAs`. 
Unfortunately, `storePubliclyAs` fails if `ACL` is not enabled.

## Scoped filesystems
I found out another bug that happens if you use `->moveFiles()` in your `FileUpload` field, as it also checks that the file in the tmp folder is on the same disk as the final destination. 
Unfortunately, this check fails if you are using [scoped filesystems](https://laravel.com/docs/11.x/filesystem#scoped-and-read-only-filesystems) as they have a different name than the disk in use

## TL;DR
You can fix it by using `->visibility('private')` or, if you actually want the file to be public, enabling ACL on your bucket

Hope this helps someone

**Reply by jvillegasd** [reply]
@TimeZHero I'm running into the same issue using R2 bucket as filesystem and app deployed using docker php-fpm + nginx, after adding this `->visibility('private')` I cannot see this fixed

**Reply by jvillegasd** [reply]
Never mind I just fixed by assigning the same value of `FILESYSTEM_DISK` to the env var `FILAMENT_FILESYSTEM_DISK` and adding my prod host to R2 CORS settings by allowing these http methods:

GET
HEAD
PUT
POST
DELETE
